v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/MigrateForTable.java

```diff
@@ -0,0 +1,132 @@
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.templates;
+
+import com.google.cloud.teleport.v2.constants.SourceDbToSpannerConstants;
+import com.google.cloud.teleport.v2.options.SourceDbToSpannerOptions;
+import com.google.cloud.teleport.v2.source.reader.ReaderImpl;
+import com.google.cloud.teleport.v2.source.reader.io.row.SourceRow;
+import com.google.cloud.teleport.v2.source.reader.io.transform.ReaderTransform;
+import com.google.cloud.teleport.v2.spanner.ddl.Ddl;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.ISchemaMapper;
+import com.google.cloud.teleport.v2.spanner.migrations.transformation.CustomTransformation;
+import com.google.cloud.teleport.v2.transformer.SourceRowToMutationDoFn;
+import com.google.cloud.teleport.v2.writer.DeadLetterQueue;
+import com.google.cloud.teleport.v2.writer.SpannerWriter;
+import java.util.Arrays;
+import org.apache.beam.repackaged.core.org.apache.commons.lang3.StringUtils;
+import org.apache.beam.sdk.coders.SerializableCoder;
+import org.apache.beam.sdk.io.gcp.spanner.MutationGroup;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerConfig;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerWriteResult;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MigrateForTable extends PTransform<PBegin, PCollection<Void>> {
+  private static final Logger LOG = LoggerFactory.getLogger(MigrateForTable.class);
+
+  private SourceDbToSpannerOptions options;
+  private SpannerConfig spannerConfig;
+  private Ddl ddl;
+  private ISchemaMapper schemaMapper;
+  private ReaderImpl reader;
+  private String shardId;
+
+  public MigrateForTable(
+      SourceDbToSpannerOptions options,
+      SpannerConfig spannerConfig,
+      Ddl ddl,
+      ISchemaMapper schemaMapper,
+      ReaderImpl reader,
+      String shardId) {
+    this.options = options;
+    this.spannerConfig = spannerConfig;
+    this.ddl = ddl;
+    this.schemaMapper = schemaMapper;
+    this.reader = reader;
+    this.shardId = StringUtils.isEmpty(shardId) ? "" : shardId;
+  }
+
+  @Override
+  public PCollection<Void> expand(PBegin input) {
+    ReaderTransform readerTransform = reader.getReaderTransform();
+
+    PCollectionTuple rowsAndTables = input.apply("Read_rows", readerTransform.readTransform());
```

Comment:
suggest we name all the stages with shard and table name or something so it is easy to relate.

Reply:
Since we have wrapped it in a ptransform now, we set the shard/table identified at the top level pipeline. Adding additional identifiers for steps will be redundant and have overhead of passing these parameters everywhere where.

All our stages now look like:
Migrate_for_<shard_id>_<table_name>/Read_rows/...
Migrate_for_<table_name>/Write_To_Spanner/... (non-sharded cases)